cellarium/ml/models/nmf.py

```diff
@@ -189,79 +477,83 @@ def solve_alpha_wKL(self, alpha_nk: torch.Tensor, x_ng: torch.Tensor, n_iteratio
-            alpha_buffer = alpha_nk.exp().clone()
+            alpha_buffer = alpha_rnk.exp().clone()
 
-        # return F.softplus(alpha_nk).detach()
-        return alpha_nk.exp().detach()
+        return alpha_rnk.exp().detach()
 
-    def dictionary_update(self, factors_kg: torch.Tensor, n_iterations: int = 1) -> torch.Tensor:
+    # @cuda.jit
+    def dictionary_update_3d(self, factors_kg: torch.Tensor, n_iterations: int = 1) -> torch.Tensor:
         """
         Algorithm 2 from Mairal et al. [1] for computing the dictionary update.
 
         Args:
             factors_kg: The matrix of gene expression programs (Mairal's dictionary D).
             n_iterations: The number of iterations to perform.
         """
+
+        i = factors_kg.shape[1]
         D_buffer = factors_kg.clone()
-        k_dimension = factors_kg.shape[0]
         updated_factors_kg = factors_kg.clone()
 
         for _ in range(n_iterations):
-            for k in range(k_dimension):
-                scalar = self.A_kk[k, k]
-                a_1k = self.A_kk[k, :]
-                b_1g = self.B_kg[k, :]
+            for k in range(i):
+
+                A_rkk = getattr(self, f"A_{i}_rkk")
+                B_rkg = getattr(self, f"B_{i}_rkg")
+
+                scalar = A_rkk[:, k, k].view(self.n_nmf, 1, 1)
+                a_1k = A_rkk[:, k, :].unsqueeze(1)
+                b_1g = B_rkg[:, k, :].unsqueeze(1)
 
                 # Algorithm 2 line 3 with added non-negativity constraint, also possibly wrong
                 u_1g = torch.clamp(
-                    updated_factors_kg[k, :] + (b_1g - torch.matmul(a_1k, updated_factors_kg)) / scalar,
+                    updated_factors_kg[:, k, :].unsqueeze(1)
+                    + (b_1g - torch.bmm(a_1k, updated_factors_kg)) / scalar,
                     min=0.0,
                 )
 
-                updated_factors_1g = u_1g / torch.clamp(torch.linalg.norm(u_1g), min=1.0)
-                updated_factors_kg[k, :] = updated_factors_1g
+                u_1g_reshape = u_1g.squeeze(1)
+                u_1g_reshape = torch.linalg.norm(u_1g_reshape, dim=1)
+
+                updated_factors_1g = u_1g / torch.clamp(u_1g_reshape.view(self.n_nmf, 1, 1), min=1.0)
+                updated_factors_kg[:, k, :] = updated_factors_1g.squeeze(1)
 
-            D_diff = torch.linalg.norm(updated_factors_kg - D_buffer) / torch.linalg.norm(updated_factors_kg)
+            D_diff = torch.linalg.norm(
+                updated_factors_kg.view(-1, self.n_vars) - D_buffer.view(-1, self.n_vars)
+            ) / torch.linalg.norm(updated_factors_kg.view(-1, self.n_vars))
             if D_diff <= self._D_tol:
                 break
             D_buffer = updated_factors_kg.clone()
 
         return updated_factors_kg
 
-    def forward(self, x_ng: torch.Tensor, var_names_g: np.ndarray) -> dict[str, torch.Tensor | None]:
+    def nmf_forward(self, x_ng: torch.Tensor, var_names_g: np.ndarray) -> dict[str, torch.Tensor | None]:
```

Comment:
I do not think we should split things into `nmf_forward()` and `consensus_forward()`.  Can't we have just one `forward()` function?  If `r == 1`, then we are doing regular NMF, and if `r > 1` then we are doing consensus NMF.  But the code could be the exact same.  Just the `r` dimension of tensors will have size 1 if we are doing regular NMF.  Right?